File: c/hash-table/maximal-rectangle.c
int maximalRectangle(char** matrix, int matrixRowSize, int matrixColSize) {
    int max_area = 0;
    int h,w;
    int *height;
    int *st;
    int i,j;
    int top;

    height = malloc(sizeof(int) * (matrixColSize + 1));
    st = malloc(sizeof(int) * (matrixColSize + 1));
    memset(height, 0, sizeof(int) * (matrixColSize + 1));

    for(i = 0; i < matrixRowSize; ++i){
        top = -1;
        for(j = 0; j <= matrixColSize; ++j){
            if(j < matrixColSize)
                height[j] = matrix[i][j] == '1' ? height[j] + 1 : 0;

            while(top > -1 && height[j] <= height[st[top]]){
                h = height[st[top--]];
                w = ((top == -1) ? j : (j - st[top] - 1));

                if(h * w > max_area)
                    max_area = h * w;
            }

            st[++top] = j;
        }
    }

    free(height);
    free(st);

    return max_area;
}

File: python/hash-table/maximal-rectangle.py
class Solution(object):
    def maximalRectangle(self, matrix):
        if len(matrix) == 0:
            return 0

        max_area = 0
        st = [0] * (len(matrix[0]) + 1)
        height = [0] * (len(matrix[0]) + 1)
        for i in range(len(matrix)):
            top = -1
            for j in range(len(matrix[i]) + 1):
                if j < len(matrix[i]):
                    height[j] = ((height[j] + 1) if (matrix[i][j] == '1') else 0)

                while top > -1 and height[j] < height[st[top]]:
                    h = height[st[top]]
                    top -= 1
                    w = j if (top == -1) else (j - st[top] - 1)
                    if h * w > max_area:
                        max_area = h * w
                top += 1
                st[top] = j

        return max_area
